### PR TITLE
Make SmallVectorImpl destructor protected

### PR DESCRIPTION
**Pull request by andykaylor**
Because the SmallVectorImpl destructor is not virtual, the destructor of derived classes will not be called if pointers to the SmallVectorImpl class are deleted directly. Making the SmallVectorImpl destructor protected will prevent this.

### CODE REVIEW

File: mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp
Comment (kuhar) [inline mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp:98]: IMO this should either return `SmallVector<char>` (without an exact size) or take `SmallVectorImpl<char> &` as the second function argument

**Review by kuhar — COMMENTED** [review]
Overall I think this makes sense, but the `SerializeToHsaco` changes look suboptimal

File: mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp
Comment (andykaylor) [reply mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp:98]: I'm not familiar enough with this code to know what would be best, and honestly I didn't spend any time trying to figure that out. My goal was just to get it to compile with my SmallVectorImpl change. I hope that the maintainers of this code will make a change that will make my change unnecessary.

Comment (joker-eph) [reply mlir/lib/Dialect/GPU/Transforms/SerializeToHsaco.cpp:98]: @krzysz00 can you change this ahead of the current PR?

**Comment by dwblaikie** [comment]
We probably (pretty sure) don't want to add a virtual dtor to SmallVector - that'd add a vtable pointer, increasing the size in ways that are probably unacceptable given the pervasive use of the data structure.

We should make the Impl dtor protected so it can't be polymorphically destroyed.

**Comment by andykaylor** [comment]
> We probably (pretty sure) don't want to add a virtual dtor to SmallVector - that'd add a vtable pointer, increasing the size in ways that are probably unacceptable given the pervasive use of the data structure.
> 
> We should make the Impl dtor protected so it can't be polymorphically destroyed.

That's what I thought as well. It's not particularly clear since GitHub is highlighting the deleted constructor as the thing that changed, but that is what this PR does -- the "public" line is dropped below the destructor, making it protected.

**Review by kuhar — APPROVED** [review]
I think we have reached consensus that this is the preferred direction. `SerializeToHsaco` is a tiny detail in the grand scheme of things, and because this PR doesn't make it any worse, I don't think we should be blocked by it.

**Comment by krzysz00** [comment]
I don't know if I'll be able to get to the SerializeToHsaco fix today, but passing in `SmallVectorImpl<char>&` would be my preferred solution ...

Or really that should be `MemoryBuffer &` or some other such structure if feasible.

**Comment by andykaylor** [comment]
I don't think this change is time-critical. I can wait a few days for the SerializeToHsaco fix, but if it's going to be more than a few days, I'd prefer to land this change with an interim fix in SerializeToHsaco. @krzysz00 what's your preference?

**Comment by krzysz00** [comment]
I put up a PR to fix SerializeToHsaco and unblock this

**Comment by andykaylor** [comment]
I'm not sure what happened here. I reset the branch in my fork so I could update this with the #71702 change in place. I'm testing locally and plan to update this PR (or create a new one) when I'm done.

**Comment by joker-eph** [comment]
You can always force-push here the same branch you pushed for the new PR: that’ll set the same state.

> I think we have reached consensus that this is the preferred direction. SerializeToHsaco is a tiny detail in the grand scheme of things, and because this PR doesn't make it any worse, I don't think we should be blocked by it.


This is solved, but to keep in mind for next time: this is the kind of change to land separately.
First because we can, but second because this PR could break something else and be reverted / relabeled a couple of times. No need to couple the changes and add churn in the other component (here MLIR). Also it can increase the chance of conflicts if someone wanted to revert (because of evolution in the other parts of the code): so just good practices :)

**Comment by andykaylor** [comment]
> This is solved, but to keep in mind for next time: this is the kind of change to land separately. First because we can, but second because this PR could break something else and be reverted / relabeled a couple of times. No need to couple the changes and add churn in the other component (here MLIR). Also it can increase the chance of conflicts if someone wanted to revert (because of evolution in the other parts of the code): so just good practices :)

Sure, that makes a lot of sense, and I didn't really expect this to land with the MLIR changes as part of my commit. I was just including them in the first revision because otherwise my change would have made things unbuildable.

**Pull request closed without merging**
